### PR TITLE
Add artifact check for between RHBQ and upstream Quarkus

### PR DESCRIPTION
**Pull request by jedla97**
This PR adding the check for missing/extra artifacts between RHBQ and upstream platform boms. In addition it checks the version of these artifacts (not thoroughly as previously developed check) so we checking it 2 times.

To check how this addition work you can use one of these commands using the RHBQ 3.8.5 maven-repo zip

`mvn clean install exec:java -DskipTests -Dquarkus.platform.bom=com.redhat.quarkus.platform:quarkus-bom:3.8.5.redhat-00003 -Dmaven.repo.local=<path to unziped RHBQ> -Dquarkus.repo.tag=3.8.5 -Dquarkus-version=3.8`

`mvn clean test -Dtest="DiffReportTest#checkForDifferentArtifactsInPlatformBoms" -Dquarkus.platform.bom=com.redhat.quarkus.platform:quarkus-bom:3.8.5.redhat-00003 -Dmaven.repo.local=<path to unziped RHBQ> -Dquarkus.repo.tag=3.8.5 -Dquarkus-version=3.8`

or for full all the checks

`mvn clean test -Dquarkus.platform.bom=com.redhat.quarkus.platform:quarkus-bom:3.8.5.redhat-00003 -Dmaven.repo.local=<path to unziped RHBQ> -Dquarkus.repo.tag=3.8.5 -Dquarkus-version=3.8`

Note that `mvn clean test` will fail as the `allowed_artifact.yaml` wasn't updated with latest exception!

### CODE REVIEW

**Comment by jedla97** [comment]
@rsvoboda can you look at this as you have most context about this additional testing

**Comment by rsvoboda** [comment]
I will check this soon, it slipped through.

File: artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+     * Generate report of platform bom comparison of artifacts
+     * The output is saved to platformBomComparison.html
+     * There is need to check if `allowedArtifactsFile` is not null otherwise the exception is thrown.
+     */
+    public void generatePomComparisonReport() {
+        LOG.info("Generating platform bom comparison report");
+        try(FileWriter fw = new FileWriter("platformBomComparison.html")) {
+            fw.write(HTML_BASE_START);
+
+            fw.write("<h1>RHBQ missing artifacts</h1>\n<table>");
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java]:
```suggestion
            fw.write("<h1>RHBQ BOM - missing artifacts</h1>\n<table>");
```

File: artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            fw.write(generateDifferArtifacts(pomComparator.getMissingDependencies(),
+                    allowedArtifactsFile != null ? allowedArtifactsFile.getBomComparisonsMissingArtifacts() : null));
+            fw.write("</table>");
+
+            fw.write("<h1>RHBQ extra artifacts</h1>\n<table>");
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java]:
```suggestion
            fw.write("<h1>RHBQ BOM - extra artifacts</h1>\n<table>");
```

File: artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+            fw.write(generateDifferArtifacts(pomComparator.getExtraDependencies(),
+                    allowedArtifactsFile != null ? allowedArtifactsFile.getBomComparisonsExtraArtifacts() : null));
+            fw.write("</table>");
+
+            fw.write("<h1>RHBQ artifacts with different version</h1>\n<table>");
+            fw.write(generateDifferentVersions(pomComparator.getMismatchedVersion(),
+                    allowedArtifactsFile != null ? allowedArtifactsFile.getVersionComparisonsArtifacts() : null));
+            fw.write("</table>");
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java]: This should be dropped as it's duplicate for report we have in `outputDiff.html`

File: artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+     * The output is saved to platformBomComparison.html
+     * There is need to check if `allowedArtifactsFile` is not null otherwise the exception is thrown.
+     */
+    public void generatePomComparisonReport() {
+        LOG.info("Generating platform bom comparison report");
+        try(FileWriter fw = new FileWriter("platformBomComparison.html")) {
+            fw.write(HTML_BASE_START);
+
+            fw.write("<h1>RHBQ missing artifacts</h1>\n<table>");
+            fw.write(generateDifferArtifacts(pomComparator.getMissingDependencies(),
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java:75]: Artifacts should be listed alphabetically, please adjust called logic.

File: artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+                    allowedArtifactsFile != null ? allowedArtifactsFile.getBomComparisonsMissingArtifacts() : null));
+            fw.write("</table>");
+
+            fw.write("<h1>RHBQ extra artifacts</h1>\n<table>");
+            fw.write(generateDifferArtifacts(pomComparator.getExtraDependencies(),
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/GeneratePomComparison.java:80]: Artifacts should be listed alphabetically, please adjust called logic.

File: artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java [hunk 6]
```diff
@@ -79,4 +85,81 @@ public static List<String> prepareMavenProperties() {
...
+        return localRepo;
+    }
+
+    /**
+     * Download the upstream platform bom and return its path
+     * @return path to downloaded platform bom
+     */
+    public static Path getUpstreamBom() {
+        LOG.info("Executing mvn dependency:get");
+        String tag = Objects.requireNonNull(System.getProperty("quarkus.repo.tag"), "The quarkus.platform.bom wasn't set.");
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java]:
```suggestion
        String version = Objects.requireNonNull(System.getProperty("quarkus.repo.tag"), "The quarkus.platform.bom wasn't set.");
```

File: artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java [hunk 7]
```diff
@@ -79,4 +85,81 @@ public static List<String> prepareMavenProperties() {
...
+
+        return Paths.get(getLocalRepository(), "io", "quarkus", "platform", "quarkus-bom", tag, "quarkus-bom-" + tag + ".pom");
+    }
+
+    /**
+     * Get RHBQ platform bom from local repository and return its path
+     * @return path to platform bom
+     */
+    public static Path getRHBQBom() {
+        String tag = Objects.requireNonNull(System.getProperty("quarkus.platform.bom"), "The quarkus.platform.bom wasn't set.");
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java]:
```suggestion
        String bom = Objects.requireNonNull(System.getProperty("quarkus.platform.bom"), "The quarkus.platform.bom wasn't set.");
```

File: artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java [hunk 8]
```diff
@@ -79,4 +85,81 @@ public static List<String> prepareMavenProperties() {
...
+        tag = tag.substring(tag.lastIndexOf(":") + 1);
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java]:
```suggestion
        version = bom.substring(tag.lastIndexOf(":") + 1);
```

File: artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java [hunk 9]
```diff
@@ -79,4 +85,81 @@ public static List<String> prepareMavenProperties() {
...
+        return Paths.get(getLocalRepository(), "com", "redhat", "quarkus", "platform", "quarkus-bom", tag, "quarkus-bom-" + tag + ".pom");
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java]:
```suggestion
        return Paths.get(getLocalRepository(), "com", "redhat", "quarkus", "platform", "quarkus-bom", version, "quarkus-bom-" + tag + ".pom");
```

File: artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java [hunk 10]
```diff
@@ -79,4 +85,81 @@ public static List<String> prepareMavenProperties() {
...
     }
 
+    /**
+     * Check for maven.repo.local property to propagated it. If the property is not set the default M2 home
+     * @return String path to local repository
+     */
+    public static String getLocalRepository() {
+        String localRepo = Objects.requireNonNullElse(System.getProperty("maven.repo.local"), "");
+        if (localRepo.isEmpty()) {
+            localRepo = "~/.m2/repository/";
```
Comment (rsvoboda) [inline artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java]: `~` is bash/shell thingy 

`System.getProperty("user.home")` should be more safe + you should consider this to be also Windows friendly

**Review by rsvoboda — CHANGES_REQUESTED** [review]

File: artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java (hunk 10 above)
Comment (jedla97) [reply artifact-version-diff/src/main/java/io/quarkus/qe/PrepareOperation.java]: Good catch! Updated it to be more safe on multiplatform.

**Comment by jedla97** [comment]
Thanks for review! All comments are updated + I added second commit containing reasoning why the items in `3.8_allowed_artifacs.yaml` were allowed to have different version and removing jna as it was fixed.

**Comment by rsvoboda** [comment]
Will need rebase + please add header about what versions are compared to the reports + a footer saying  "generated  on date + time"

**Comment by jedla97** [comment]
Rebased, conflicts resolved, and added title + footer

**Review by rsvoboda — APPROVED** [review]

**Comment by rsvoboda** [comment]
@jedla97 don't forget to update description of utils-compare-artifacts-upstream-vs-prod to point also to the new platformBomComparison.html report

**Pull request merged**